File: apps/desktop/src/lib/projectState.ts
import fixtures from "../../fixtures/project-state.json";
import { z } from "zod";

export const PROJECT_STATE_SCHEMA_VERSION = 1 as const;

const identifierSchema = z
  .string()
  .min(1)
  .max(96)
  .regex(/^[a-z0-9._/-]+$/u);
const commitSchema = z.string().regex(/^[0-9a-f]{40}$/u);
const nullableText = z.string().min(1).max(4096).nullable();
const optionalTimestamp = z.string().datetime().nullable();

export const provenanceSchema = z
  .object({
    trust: z.enum(["verified", "reported", "inferred", "unknown"]),
    sourceType: z.enum([
      "git",
      "repository-document",
      "validation-report",
      "package-manifest",
      "application-storage",
      "agent-session",
      "user-approval",
      "unknown",
    ]),
    sourceRef: nullableText,
    sourceCommit: commitSchema.nullable(),
    observedAt: optionalTimestamp,
    verifiedAt: optionalTimestamp,
    note: nullableText,
  })
  .strict();

const approvalSchema = z
  .object({
    decision: z.enum([
      "required",
      "approved",
      "rejected",
      "superseded",
      "unknown",
    ]),
    authority: nullableText,
    approvedAt: optionalTimestamp,
    scope: nullableText,
    supersededAt: optionalTimestamp,
    provenance: provenanceSchema,
  })
  .strict()
  .superRefine((approval, context) => {
    if (
      ["approved", "rejected", "superseded"].includes(approval.decision) &&
      approval.authority === null
    ) {
      context.addIssue({
        code: "custom",
        message: "A decided approval requires an authority",
      });
    }
  });

const checkpointStatusSchema = z.enum(["pushed", "paused", "finished"]);
const validationResultSchema = z.enum([
  "passed",
  "failed",
  "blocked",
  "not-run",
]);

export const checkpointStateSchema = z
  .object({
    status: checkpointStatusSchema,
    commit: commitSchema.nullable(),
    branch: nullableText,
    pushed: z.boolean(),
    validationsCurrent: z.boolean(),
    documentationCurrent: z.boolean(),
    completionClaimed: z.boolean(),
    timestamp: optionalTimestamp,
    remainingWork: z.array(z.string().min(1).max(4096)).max(256),
    provenance: provenanceSchema,
  })
  .strict()
  .superRefine((checkpoint, context) => {
    const valid =
      (checkpoint.status === "pushed" &&
        checkpoint.pushed &&
        checkpoint.commit !== null &&
        !checkpoint.completionClaimed) ||
      (checkpoint.status === "paused" && !checkpoint.completionClaimed) ||
      (checkpoint.status === "finished" &&
        checkpoint.pushed &&
        checkpoint.commit !== null &&
        checkpoint.completionClaimed &&
        checkpoint.remainingWork.length === 0);
    if (!valid) {
      context.addIssue({
        code: "custom",
        message: "Checkpoint fields are inconsistent",
      });
    }
  });

const workSessionSchema = z
  .object({
    id: identifierSchema,
    actor: z.string().min(1).max(128),
    executionContext: z.enum(["local", "remote", "unknown"]),
    status: z.enum(["active", "paused", "complete", "unknown"]),
    target: z.string().min(1).max(4096),
    startedAt: optionalTimestamp,
    endedAt: optionalTimestamp,
    lastActivityAt: optionalTimestamp,
    pauseReason: nullableText,
    uncommittedWorkMayExist: z.boolean(),
    provenance: provenanceSchema,
  })
  .strict();

const validationStateSchema = z
  .object({
    category: identifierSchema,
    checkId: identifierSchema,
    command: nullableText,
    result: validationResultSchema,
    scope: z.string().min(1).max(4096),
    timestamp: optionalTimestamp,
    evidenceRef: nullableText,
    commitTested: commitSchema.nullable(),
    current: z.boolean(),
    blockerId: identifierSchema.nullable(),
    provenance: provenanceSchema,
  })
  .strict();

const packageRequirementsSchema = z
  .object({
    required: z.boolean(),
    evidence: z
      .array(
        z
          .object({
            artifactType: identifierSchema,
            path: nullableText,
            filename: nullableText,
            sourceCommit: commitSchema.nullable(),
            sizeBytes: z.number().int().nonnegative().safe().nullable(),
            sha256: z
              .string()
              .regex(/^[0-9a-f]{64}$/u)
              .nullable(),
            manifestRef: nullableText,
            platformBaseline: nullableText,
            installResult: validationResultSchema.nullable(),
            launchResult: validationResultSchema.nullable(),
            desktopIntegrationResult: validationResultSchema.nullable(),
            smokeTestResult: validationResultSchema.nullable(),
            provenance: provenanceSchema,
          })
          .strict(),
      )
      .max(32),
    provenance: provenanceSchema,
  })
  .strict();

const boundariesSchema = z
  .object({
    approvedActions: z.array(z.string().min(1).max(4096)).max(256),
    prohibitedActions: z.array(z.string().min(1).max(4096)).max(256),
    confirmationRequiredActions: z.array(z.string().min(1).max(4096)).max(256),
    approvals: z.array(approvalSchema).max(256),
    provenance: provenanceSchema,
  })
  .strict();

const blockerSchema = z
  .object({
    id: identifierSchema,
    description: z.string().min(1).max(4096),
    severity: z.enum(["info", "warning", "error"]),
    affectedRequirement: z.string().min(1).max(4096),
    external: z.boolean(),
    preExisting: z.boolean(),
    milestoneCaused: z.boolean(),
    recommendedAction: z.string().min(1).max(4096),
    approvalRequired: z.boolean(),
    provenance: provenanceSchema,
  })
  .strict();

const contradictionSchema = z
  .object({
    id: identifierSchema,
    kind: identifierSchema,
    description: z.string().min(1).max(4096),
    affectedRequirement: nullableText,
    provenance: provenanceSchema,
  })
  .strict();

export const projectStateContractSchema = z
  .object({
    schemaVersion: z.literal(PROJECT_STATE_SCHEMA_VERSION),
    project: z
      .object({
        id: identifierSchema,
        displayName: z.string().min(1).max(120),
        repository: z.string().regex(/^[a-z0-9._-]+\/[a-z0-9._-]+$/u),
        localWorkspaceId: identifierSchema.nullable(),
        primaryPlatform: identifierSchema,
        activeUiPlatform: identifierSchema,
        productDirectionRef: z.string().min(1).max(4096),
        lifecyclePhase: identifierSchema,
        provenance: provenanceSchema,
      })
      .strict(),
    roadmapRef: z.string().min(1).max(4096),
    repository: z
      .object({
        currentBranch: nullableText,
        baseBranch: nullableText,
        localHead: commitSchema.nullable(),
        remoteHead: commitSchema.nullable(),
        ahead: z.number().int().nonnegative().safe().nullable(),
        behind: z.number().int().nonnegative().safe().nullable(),
        worktree: z.enum(["clean", "dirty", "unknown"]),
        lastVerifiedCheckpoint: commitSchema.nullable(),
        mergeAuthorization: approvalSchema,
        releaseAuthorization: approvalSchema,
        provenance: provenanceSchema,
      })
      .strict()
      .superRefine((repository, context) => {
        if (
          (repository.ahead === null) !== (repository.behind === null) ||
          (repository.localHead === null) !== (repository.remoteHead === null)
        ) {
          context.addIssue({
            code: "custom",
            message: "Repository pairs are inconsistent",
          });
        }
      }),
    milestone: z
      .object({
        id: identifierSchema,
        title: z.string().min(1).max(256),
        status: z.enum(["planned", "active", "paused", "complete", "blocked"]),
        objective: z.string().min(1).max(4096),
        approvedScope: z.array(z.string().min(1).max(4096)).max(256),
        exclusions: z.array(z.string().min(1).max(4096)).max(256),
        completionRequirements: z.array(z.string().min(1).max(4096)).max(256),
        predecessorId: identifierSchema.nullable(),
        successorId: identifierSchema.nullable(),
        ownerApproval: approvalSchema,
        provenance: provenanceSchema,
      })
      .strict(),
    workSessions: z.array(workSessionSchema).max(256),
    checkpoints: z.array(checkpointStateSchema).max(256),
    validations: z.array(validationStateSchema).max(1024),
    packages: packageRequirementsSchema,
    boundaries: boundariesSchema,
    blockers: z.array(blockerSchema).max(256),
    contradictions: z.array(contradictionSchema).max(256),
    nextAction: z
      .object({
        action: z.string().min(1).max(4096),
        why: z.string().min(1).max(4096),
        approvalRequired: z.boolean(),
        targetMilestone: identifierSchema.nullable(),
        requiredStartingCommit: commitSchema.nullable(),
        requiredBranch: nullableText,
        provenance: provenanceSchema,
      })
      .strict()
      .nullable(),
    handoff: z
      .object({
        status: checkpointStatusSchema,
        phrase: z.string().min(1).max(256),
        generatedAt: optionalTimestamp,
        sourceCheckpoint: commitSchema.nullable(),
        provenance: provenanceSchema,
      })
      .strict()
      .superRefine((handoff, context) => {
        const expected = {
          pushed: "Codex checkpoint pushed. Continue.",
          paused: "Codex paused. Continue.",
          finished: "Codex finished. Continue.",
        }[handoff.status];
        if (handoff.phrase !== expected) {
          context.addIssue({
            code: "custom",
            message: "Handoff phrase is inconsistent",
          });
        }
      }),
    provenance: provenanceSchema,
  })
  .strict();

export type ProjectStateContract = z.infer<typeof projectStateContractSchema>;
export const projectStateFixturesSchema = z
  .object({
    minimalValid: projectStateContractSchema,
    activeMilestone: projectStateContractSchema,
    pushedCheckpoint: checkpointStateSchema,
    pausedSession: workSessionSchema,
    completedMilestone: checkpointStateSchema,
    missingEvidence: packageRequirementsSchema,
    missingValidationEvidence: validationStateSchema,
    contradictoryEvidence: z.array(contradictionSchema).max(256),
  })
  .strict();
export const projectStateFixtures = projectStateFixturesSchema.parse(fixtures);
export const scaffoldProjectState = projectStateContractSchema.parse(
  projectStateFixtures.activeMilestone,
);

File: apps/desktop/src-tauri/src/project_state.rs
use std::collections::HashSet;

use serde::{Deserialize, Serialize};
use thiserror::Error;

pub const PROJECT_STATE_SCHEMA_VERSION: u16 = 1;

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum TrustClassification {
    Verified,
    Reported,
    Inferred,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum StateSourceType {
    Git,
    RepositoryDocument,
    ValidationReport,
    PackageManifest,
    ApplicationStorage,
    AgentSession,
    UserApproval,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Provenance {
    pub trust: TrustClassification,
    pub source_type: StateSourceType,
    pub source_ref: Option<String>,
    pub source_commit: Option<String>,
    pub observed_at: Option<String>,
    pub verified_at: Option<String>,
    pub note: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ApprovalDecision {
    Required,
    Approved,
    Rejected,
    Superseded,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ApprovalState {
    pub decision: ApprovalDecision,
    pub authority: Option<String>,
    pub approved_at: Option<String>,
    pub scope: Option<String>,
    pub superseded_at: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProjectIdentity {
    pub id: String,
    pub display_name: String,
    pub repository: String,
    pub local_workspace_id: Option<String>,
    pub primary_platform: String,
    pub active_ui_platform: String,
    pub product_direction_ref: String,
    pub lifecycle_phase: String,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum WorktreeState {
    Clean,
    Dirty,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RepositoryState {
    pub current_branch: Option<String>,
    pub base_branch: Option<String>,
    pub local_head: Option<String>,
    pub remote_head: Option<String>,
    pub ahead: Option<u32>,
    pub behind: Option<u32>,
    pub worktree: WorktreeState,
    pub last_verified_checkpoint: Option<String>,
    pub merge_authorization: ApprovalState,
    pub release_authorization: ApprovalState,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum MilestoneStatus {
    Planned,
    Active,
    Paused,
    Complete,
    Blocked,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MilestoneState {
    pub id: String,
    pub title: String,
    pub status: MilestoneStatus,
    pub objective: String,
    pub approved_scope: Vec<String>,
    pub exclusions: Vec<String>,
    pub completion_requirements: Vec<String>,
    pub predecessor_id: Option<String>,
    pub successor_id: Option<String>,
    pub owner_approval: ApprovalState,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ExecutionContext {
    Local,
    Remote,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum SessionStatus {
    Active,
    Paused,
    Complete,
    Unknown,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WorkSessionState {
    pub id: String,
    pub actor: String,
    pub execution_context: ExecutionContext,
    pub status: SessionStatus,
    pub target: String,
    pub started_at: Option<String>,
    pub ended_at: Option<String>,
    pub last_activity_at: Option<String>,
    pub pause_reason: Option<String>,
    pub uncommitted_work_may_exist: bool,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum CheckpointStatus {
    Pushed,
    Paused,
    Finished,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CheckpointState {
    pub status: CheckpointStatus,
    pub commit: Option<String>,
    pub branch: Option<String>,
    pub pushed: bool,
    pub validations_current: bool,
    pub documentation_current: bool,
    pub completion_claimed: bool,
    pub timestamp: Option<String>,
    pub remaining_work: Vec<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ValidationResult {
    Passed,
    Failed,
    Blocked,
    NotRun,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ValidationState {
    pub category: String,
    pub check_id: String,
    pub command: Option<String>,
    pub result: ValidationResult,
    pub scope: String,
    pub timestamp: Option<String>,
    pub evidence_ref: Option<String>,
    pub commit_tested: Option<String>,
    pub current: bool,
    pub blocker_id: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PackageEvidence {
    pub artifact_type: String,
    pub path: Option<String>,
    pub filename: Option<String>,
    pub source_commit: Option<String>,
    pub size_bytes: Option<u64>,
    pub sha256: Option<String>,
    pub manifest_ref: Option<String>,
    pub platform_baseline: Option<String>,
    pub install_result: Option<ValidationResult>,
    pub launch_result: Option<ValidationResult>,
    pub desktop_integration_result: Option<ValidationResult>,
    pub smoke_test_result: Option<ValidationResult>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PackageRequirements {
    pub required: bool,
    pub evidence: Vec<PackageEvidence>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Boundaries {
    pub approved_actions: Vec<String>,
    pub prohibited_actions: Vec<String>,
    pub confirmation_required_actions: Vec<String>,
    pub approvals: Vec<ApprovalState>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum BlockerSeverity {
    Info,
    Warning,
    Error,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Blocker {
    pub id: String,
    pub description: String,
    pub severity: BlockerSeverity,
    pub affected_requirement: String,
    pub external: bool,
    pub pre_existing: bool,
    pub milestone_caused: bool,
    pub recommended_action: String,
    pub approval_required: bool,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Contradiction {
    pub id: String,
    pub kind: String,
    pub description: String,
    pub affected_requirement: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct NextAction {
    pub action: String,
    pub why: String,
    pub approval_required: bool,
    pub target_milestone: Option<String>,
    pub required_starting_commit: Option<String>,
    pub required_branch: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct HandoffState {
    pub status: CheckpointStatus,
    pub phrase: String,
    pub generated_at: Option<String>,
    pub source_checkpoint: Option<String>,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProjectStateContract {
    pub schema_version: u16,
    pub project: ProjectIdentity,
    pub roadmap_ref: String,
    pub repository: RepositoryState,
    pub milestone: MilestoneState,
    pub work_sessions: Vec<WorkSessionState>,
    pub checkpoints: Vec<CheckpointState>,
    pub validations: Vec<ValidationState>,
    pub packages: PackageRequirements,
    pub boundaries: Boundaries,
    pub blockers: Vec<Blocker>,
    pub contradictions: Vec<Contradiction>,
    pub next_action: Option<NextAction>,
    pub handoff: HandoffState,
    pub provenance: Provenance,
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
pub enum ProjectStateContractError {
    #[error("project-state schema version is unsupported")]
    UnsupportedVersion,
    #[error("project-state contract contains an invalid identifier")]
    InvalidIdentifier,
    #[error("project-state contract contains duplicate identities")]
    DuplicateIdentity,
    #[error("project-state contract has an invalid checkpoint state")]
    InvalidCheckpoint,
    #[error("project-state contract has an invalid approval state")]
    InvalidApproval,
    #[error("project-state contract has inconsistent repository state")]
    InvalidRepository,
}

impl ProjectStateContract {
    pub fn validate(&self) -> Result<(), ProjectStateContractError> {
        if self.schema_version != PROJECT_STATE_SCHEMA_VERSION {
            return Err(ProjectStateContractError::UnsupportedVersion);
        }
        if !is_identifier(&self.project.id)
            || !is_repository(&self.project.repository)
            || !is_identifier(&self.milestone.id)
            || self.roadmap_ref.is_empty()
        {
            return Err(ProjectStateContractError::InvalidIdentifier);
        }
        validate_approval(&self.milestone.owner_approval)?;
        validate_approval(&self.repository.merge_authorization)?;
        validate_approval(&self.repository.release_authorization)?;
        for approval in &self.boundaries.approvals {
            validate_approval(approval)?;
        }
        if self.repository.ahead.is_some() != self.repository.behind.is_some() {
            return Err(ProjectStateContractError::InvalidRepository);
        }
        if let Some(commit) = &self.repository.local_head {
            validate_sha(commit)?;
        }
        if let Some(commit) = &self.repository.remote_head {
            validate_sha(commit)?;
        }
        let mut checkpoint_commits = HashSet::new();
        for checkpoint in &self.checkpoints {
            validate_checkpoint(checkpoint)?;
            if let Some(commit) = checkpoint.commit.as_deref() {
                if !checkpoint_commits.insert(commit) {
                    return Err(ProjectStateContractError::DuplicateIdentity);
                }
            }
        }
        for session in &self.work_sessions {
            if !is_identifier(&session.id) {
                return Err(ProjectStateContractError::InvalidIdentifier);
            }
        }
        for validation in &self.validations {
            if !is_identifier(&validation.check_id) {
                return Err(ProjectStateContractError::InvalidIdentifier);
            }
            if let Some(commit) = &validation.commit_tested {
                validate_sha(commit)?;
            }
        }
        for evidence in &self.packages.evidence {
            if let Some(commit) = &evidence.source_commit {
                validate_sha(commit)?;
            }
            if let Some(hash) = &evidence.sha256 {
                if hash.len() != 64
                    || !hash
                        .bytes()
                        .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit())
                {
                    return Err(ProjectStateContractError::InvalidIdentifier);
                }
            }
        }
        Ok(())
    }
}

fn validate_checkpoint(checkpoint: &CheckpointState) -> Result<(), ProjectStateContractError> {
    if let Some(commit) = &checkpoint.commit {
        validate_sha(commit)?;
    }
    let valid = match checkpoint.status {
        CheckpointStatus::Pushed => {
            checkpoint.pushed && checkpoint.commit.is_some() && !checkpoint.completion_claimed
        }
        CheckpointStatus::Paused => !checkpoint.completion_claimed,
        CheckpointStatus::Finished => {
            checkpoint.pushed
                && checkpoint.commit.is_some()
                && checkpoint.completion_claimed
                && checkpoint.remaining_work.is_empty()
        }
    };
    if valid {
        Ok(())
    } else {
        Err(ProjectStateContractError::InvalidCheckpoint)
    }
}

fn validate_approval(approval: &ApprovalState) -> Result<(), ProjectStateContractError> {
    let requires_authority = matches!(
        approval.decision,
        ApprovalDecision::Approved | ApprovalDecision::Rejected | ApprovalDecision::Superseded
    );
    if requires_authority
        && approval
            .authority
            .as_deref()
            .filter(|value| !value.is_empty())
            .is_none()
    {
        return Err(ProjectStateContractError::InvalidApproval);
    }
    Ok(())
}

fn validate_sha(value: &str) -> Result<(), ProjectStateContractError> {
    if value.len() == 40
        && value
            .bytes()
            .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit())
    {
        Ok(())
    } else {
        Err(ProjectStateContractError::InvalidIdentifier)
    }
}

fn is_identifier(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 96
        && value.bytes().all(|byte| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || matches!(byte, b'-' | b'.' | b'_' | b'/')
        })
}

fn is_repository(value: &str) -> bool {
    let mut parts = value.split('/');
    matches!((parts.next(), parts.next(), parts.next()), (Some(owner), Some(name), None) if is_identifier(owner) && is_identifier(name))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase", deny_unknown_fields)]
    struct Fixtures {
        minimal_valid: ProjectStateContract,
        active_milestone: ProjectStateContract,
        pushed_checkpoint: CheckpointState,
        paused_session: WorkSessionState,
        completed_milestone: CheckpointState,
        missing_evidence: PackageRequirements,
        missing_validation_evidence: ValidationState,
        contradictory_evidence: Vec<Contradiction>,
    }

    fn fixtures() -> Fixtures {
        serde_json::from_str(include_str!("../../fixtures/project-state.json"))
            .expect("project-state fixtures must deserialize")
    }

    #[test]
    fn shared_fixtures_are_valid_and_round_trip() {
        let fixtures = fixtures();
        let mut pushed = fixtures.active_milestone.clone();
        pushed.checkpoints = vec![fixtures.pushed_checkpoint];
        let mut paused = fixtures.active_milestone.clone();
        paused.work_sessions = vec![fixtures.paused_session];
        let mut completed = fixtures.active_milestone.clone();
        completed.checkpoints = vec![fixtures.completed_milestone];
        completed.milestone.status = MilestoneStatus::Complete;
        let mut missing = fixtures.active_milestone.clone();
        missing.packages = fixtures.missing_evidence;
        let mut missing_validation = fixtures.active_milestone.clone();
        missing_validation.validations = vec![fixtures.missing_validation_evidence];
        let mut contradictory = fixtures.active_milestone.clone();
        contradictory.contradictions = fixtures.contradictory_evidence;
        for state in [
            fixtures.minimal_valid,
            fixtures.active_milestone,
            pushed,
            paused,
            completed,
            missing,
            missing_validation,
            contradictory,
        ] {
            state.validate().expect("fixture must validate");
            let encoded = serde_json::to_string(&state).expect("state must serialize");
            let decoded: ProjectStateContract =
                serde_json::from_str(&encoded).expect("state must deserialize");
            assert_eq!(state, decoded);
        }
    }

    #[test]
    fn rejects_future_versions_bad_ids_and_invalid_completion() {
        let mut state = fixtures().active_milestone;
        state.schema_version = 2;
        assert_eq!(
            state.validate(),
            Err(ProjectStateContractError::UnsupportedVersion)
        );
        state.schema_version = 1;
        state.project.id = "invalid id".to_owned();
        assert_eq!(
            state.validate(),
            Err(ProjectStateContractError::InvalidIdentifier)
        );
        state.project.id = "quireforge".to_owned();
        state.checkpoints[0].status = CheckpointStatus::Finished;
        assert_eq!(
            state.validate(),
            Err(ProjectStateContractError::InvalidCheckpoint)
        );
    }

    #[test]
    fn rejects_unknown_fields_and_approvals_without_authority() {
        let raw = include_str!("../../fixtures/project-state.json").replace(
            "\"schemaVersion\": 1",
            "\"schemaVersion\": 1, \"unknown\": true",
        );
        assert!(serde_json::from_str::<Fixtures>(&raw).is_err());
        let invalid_trust = include_str!("../../fixtures/project-state.json")
            .replace("\"trust\": \"verified\"", "\"trust\": \"unsupported\"");
        assert!(serde_json::from_str::<Fixtures>(&invalid_trust).is_err());
        let mut state = fixtures().active_milestone;
        state.milestone.owner_approval.decision = ApprovalDecision::Approved;
        state.milestone.owner_approval.authority = None;
        assert_eq!(
            state.validate(),
            Err(ProjectStateContractError::InvalidApproval)
        );
    }
}
